scikit-learn/scikit-learn
# [MRG] Estimator tags

Fixes #6599, #7737, #6554, #6715

Also see #6510

Todo
* [x] remove fixme's, enable all tests again
* [x] rename tags so they are false by default
* [x] rewrite tagging mechanism
* [ ] rethink _required_parameters

For another PR:
* add tests on compound estimators like pipeline, gridsearchcv, ...
* fix MultiOutputEstimator
* replace _estimator_type with tags
* tag for sample weight support: has_fit_parameter doesn't work for meta-estimators
* remove cross_decomposition special cases (maybe)
* remove positive data special cases
* input type for sparse data?
* binary only classifiers
* biclustering
* make sure that enough tags are defined? ChainClassifier had no ClassifierMixin :-/

I decided to leave the ``_estimator_type`` for later because this PR is already too big.

Fun issue I just thought about: Can we even define the tags for a pipeline?
How do we know if a pipeline can handle missing data? Because the scalers pass it through (and the tag says they "handle" missing data) and the imputers impute it. So either these need different tags or we just punt on defining tags for pipelines and if you want to run a pipeline the author of the pipeline needs to set the tags specifically for that pipeline if they want to run the tests. That seems not the best, though.
